File: 1396-design-underground-system/1396-design-underground-system.js

var UndergroundSystem = function() {
    this.checkedIn=new Map();
    this.stationData=new Map();
};

/** 
 * @param {number} id 
 * @param {string} stationName 
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkIn = function(id, stationName, t) {
    this.checkedIn.set(id, [stationName, t])
};

/** 
 * @param {number} id 
 * @param {string} stationName 
 * @param {number} t
 * @return {void}
 */
UndergroundSystem.prototype.checkOut = function(id, stationName, t) {
    let [startStation, checkedInTime]=this.checkedIn.get(id)
    let travelTime=t-checkedInTime
    let stations=`${startStation}~${stationName}`
    
    if(this.stationData.has(stations)){
        let arr=this.stationData.get(stations)
        arr.push(travelTime)
    }
    else
        this.stationData.set(stations, [travelTime])
    
    this.checkedIn.delete(id)
    
};

/** 
 * @param {string} startStation 
 * @param {string} endStation
 * @return {number}
 */
UndergroundSystem.prototype.getAverageTime = function(startStation, endStation) {
    let stations=`${startStation}~${endStation}`
    let arr=this.stationData.get(stations)
    let sum=arr.reduce((total, i)=>total+i,0)
    return sum/arr.length
    
};

/** 
 * Your UndergroundSystem object will be instantiated and called as such:
 * var obj = new UndergroundSystem()
 * obj.checkIn(id,stationName,t)
 * obj.checkOut(id,stationName,t)
 * var param_3 = obj.getAverageTime(startStation,endStation)
 */

File: 1396-design-underground-system/1396-design-underground-system.py
class UndergroundSystem:

    def __init__(self):
        self.checkedIn={}
        self.stationData={}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.checkedIn[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        startStation, checkedInTime=self.checkedIn[id]
        travelTime=t-checkedInTime
        key=(startStation, stationName)

        totalTime, totalCount = self.stationData.get(key, (0, 0))
        self.stationData[key] = (totalTime+travelTime, totalCount+1)
        self.checkedIn.pop(id)
       

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        key=(startStation, endStation)
        totalTime, totalCount = self.stationData[key]
        return totalTime/totalCount


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)